File: src/Entity/Coupon.php
<?php

namespace App\Entity;

use App\Repository\CouponRepository;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity(repositoryClass: CouponRepository::class)]
#[ORM\Table(name: 'coupon')]
#[ORM\UniqueConstraint(name: 'uniq_coupon_code', columns: ['code'])]
class Coupon
{
    public const TYPE_PERCENTAGE = 'percentage';
    public const TYPE_FIXED = 'fixed';

    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column]
    private ?int $id = null;

    #[ORM\Column(length: 64, unique: true)]
    #[Assert\NotBlank]
    private string $code = '';

    #[ORM\Column(length: 20)]
    #[Assert\Choice(choices: [self::TYPE_PERCENTAGE, self::TYPE_FIXED])]
    private string $type = self::TYPE_PERCENTAGE;

    #[ORM\Column(type: 'decimal', precision: 10, scale: 2)]
    #[Assert\Positive]
    private string $valeur = '0.00';

    #[ORM\Column(name: 'date_expiration', nullable: true)]
    private ?\DateTimeImmutable $dateExpiration = null;

    #[ORM\Column(name: 'usage_max', options: ['default' => 1])]
    private int $usageMax = 1;

    #[ORM\Column(name: 'usage_count', options: ['default' => 0])]
    private int $usageCount = 0;

    #[ORM\Column(options: ['default' => true])]
    private bool $actif = true;

    #[ORM\Column(name: 'montant_minimum_panier', type: 'decimal', precision: 10, scale: 2, options: ['default' => 0])]
    private string $montantMinimumPanier = '0.00';

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getCode(): string
    {
        return $this->code;
    }

    public function setCode(string $code): static
    {
        $this->code = strtoupper(trim($code));
        return $this;
    }

    public function getType(): string
    {
        return $this->type;
    }

    public function setType(string $type): static
    {
        $this->type = $type;
        return $this;
    }

    public function getValeur(): float
    {
        return $this->valeur;
    }

    public function setValeur(float $valeur): static
    {
        $this->valeur = $valeur;
        return $this;
    }

    public function getDateExpiration(): ?\DateTimeImmutable
    {
        return $this->dateExpiration;
    }

    public function setDateExpiration(?\DateTimeImmutable $dateExpiration): static
    {
        $this->dateExpiration = $dateExpiration;
        return $this;
    }

    public function getUsageMax(): int
    {
        return $this->usageMax;
    }

    public function setUsageMax(int $usageMax): static
    {
        $this->usageMax = max(1, $usageMax);
        return $this;
    }

    public function getUsageCount(): int
    {
        return $this->usageCount;
    }

    public function setUsageCount(int $usageCount): static
    {
        $this->usageCount = max(0, $usageCount);
        return $this;
    }

    public function incrementUsage(): static
    {
        $this->usageCount++;
        return $this;
    }

    public function isActif(): bool
    {
        return $this->actif;
    }

    public function setActif(bool $actif): static
    {
        $this->actif = $actif;
        return $this;
    }

    public function getMontantMinimumPanier(): float
    {
        return $this->montantMinimumPanier;
    }

    public function setMontantMinimumPanier(float $montantMinimumPanier): static
    {
        $this->montantMinimumPanier = max(0.0, $montantMinimumPanier);
        return $this;
    }
}

File: src/main/java/org/example/model/Coupon.java
package org.example.model;

import java.time.LocalDate;

public class Coupon {
    public static final String TYPE_PERCENTAGE = "percentage";
    public static final String TYPE_FIXED = "fixed";

    private int id;
    private String code;
    private double reduction;
    private boolean isPercentage;
    /** Type texte pour l'admin (voir constantes ci-dessus). */
    private String type = TYPE_PERCENTAGE;
    private LocalDate dateExpiration;
    private boolean actif;
    private int usageCount;
    private int usageMax = 1;
    private double montantMinimumPanier;

    public Coupon() {}

    public Coupon(int id, String code, double reduction, boolean isPercentage, LocalDate dateExpiration, boolean actif) {
        this.id = id;
        this.code = code;
        this.reduction = reduction;
        this.isPercentage = isPercentage;
        this.type = isPercentage ? TYPE_PERCENTAGE : TYPE_FIXED;
        this.dateExpiration = dateExpiration;
        this.actif = actif;
    }

    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public String getCode() { return code; }
    public void setCode(String code) { this.code = code; }

    public double getReduction() { return reduction; }
    public void setReduction(double reduction) { this.reduction = reduction; }

    /** Alias UI : même valeur que {@link #reduction}. */
    public double getValeur() { return reduction; }

    public void setValeur(double valeur) {
        this.reduction = valeur;
    }

    public boolean isPercentage() { return isPercentage; }

    public void setPercentage(boolean percentage) {
        this.isPercentage = percentage;
        if (type == null || type.isBlank()) {
            this.type = percentage ? TYPE_PERCENTAGE : TYPE_FIXED;
        }
    }

    public String getType() {
        return type != null ? type : (isPercentage ? TYPE_PERCENTAGE : TYPE_FIXED);
    }

    public void setType(String type) {
        this.type = type;
        this.isPercentage = TYPE_PERCENTAGE.equalsIgnoreCase(type);
    }

    public LocalDate getDateExpiration() { return dateExpiration; }
    public void setDateExpiration(LocalDate dateExpiration) { this.dateExpiration = dateExpiration; }

    /** Pour les colonnes JavaFX PropertyValueFactory("actif"). */
    public boolean getActif() { return actif; }

    public boolean isActif() { return actif; }
    public void setActif(boolean actif) { this.actif = actif; }

    public int getUsageCount() { return usageCount; }
    public void setUsageCount(int usageCount) { this.usageCount = usageCount; }

    public int getUsageMax() { return usageMax; }
    public void setUsageMax(int usageMax) { this.usageMax = usageMax; }

    public double getMontantMinimumPanier() { return montantMinimumPanier; }
    public void setMontantMinimumPanier(double montantMinimumPanier) { this.montantMinimumPanier = montantMinimumPanier; }

    public boolean isValide() {
        if (!actif) {
            return false;
        }
        if (usageMax > 0 && usageCount >= usageMax) {
            return false;
        }
        if (dateExpiration == null) {
            return true;
        }
        return !LocalDate.now().isAfter(dateExpiration);
    }

    @Override
    public String toString() {
        return code + " - " + reduction + (isPercentage ? "%" : " DT");
    }
}
